table/block_based_table_reader.cc

```diff
@@ -2077,74 +2102,211 @@ void BlockBasedTableIterator::InitDataBlock() {
+    while (!data_block_iter_.Valid()) {
+      if (!data_block_iter_.status().ok()) {
+        return;
+      }
+
+      ResetDataIter();
+      index_iter_->Prev();
+
+      if (index_iter_->Valid()) {
+        InitDataBlock();
+        data_block_iter_.SeekToLast();
+      } else {
+        return;
+      }
+    }
+
+    if (!Valid()) {
       return;
     }
 
-    ResetDataIter();
-    index_iter_->Prev();
+    // We could have check lower bound here too, but we opt not to do it for
+    // code simplicity.
 
-    if (index_iter_->Valid()) {
-      InitDataBlock();
-      data_block_iter_.SeekToLast();
-    } else {
+    // Check to see if data_block_iter_.key() is covered by the current range
+    // tombstone. Note that range_tombstone_ is not a raw range tombstone
+    // returned from RangeDelAggregator, but a cooked one. See
+    // InitRangeTombstone().
+    if (range_tombstone_.start_key_.empty()) {
+      // The range tombstone extends to the beginning of the sstable.
+      if (range_tombstone_.seq_ == 0) {
+        // The range doesn't apply to the keys in the sstable. Return the
+        // entry.
+        return;
+      }
+      // The range tombstone is a deletion tombstone. Loop to allow moving to
+      // the previous file.
+      ResetDataIter();
+      return;
+    }
+
+    auto ukey = user_key();
+    if (icomp_.user_comparator()->Compare(
+            ukey, range_tombstone_.start_key_) < 0) {
+      // The key is past the tombstone. Grab the tombstone covering the
+      // current key. The new tombstone might cover the existing key, so loop
+      // so that we can have the proper check for whether the tombstone covers
+      // the key and is a deletion tombstone or not.
+      InitRangeTombstone(ukey);
+      continue;
+    }
+    // The key is contained within the current tombstone.
+    if (range_tombstone_.seq_ == 0) {
+      // The tombstone doesn't apply to the sstable. Return the entry.
       return;
     }
+
+    // Find the previous non-deleted key.
+    index_iter_->Seek(tombstone_internal_start_key());
+    if (!index_iter_->Valid()) {
+      index_iter_->SeekToLast();
+      if (!index_iter_->Valid()) {
+        ResetDataIter();
+        block_iter_points_to_real_block_ = false;
+        return;
+      }
+    }
+    InitDataBlock();
+    data_block_iter_.SeekForPrev(tombstone_internal_end_key());
+    if (Valid()) {
+      InitRangeTombstone(user_key());
+    }
   }
+}
 
-  // We could have check lower bound here too, but we opt not to do it for
-  // code simplicity.
+void BlockBasedTableIterator::InitRangeTombstone(const Slice& target) {
+  if (range_del_agg_ == nullptr || file_meta_ == nullptr) {
+    return;
+  }
+
+  range_tombstone_ = range_del_agg_->GetTombstone(target, file_meta_->largest_seqno);
+  // Clear the start key if it is less than the smallest key in the
+  // sstable. This allows us to avoid comparisons during Prev() in the common
+  // case.
+  if (!range_tombstone_.start_key_.empty() &&
```

Comment:
Oh, good point. Thanks for catching this. Yeah, a couple of instance variables will be needed.

I'm slammed for the rest of the day and tomorrow. I'll address this and the other feedback and rebase on Friday.

Reply:
I'm having trouble coming up with a test case that fails because of this. If `range_tombstone_.start_key_` is empty, either we don't have a tombstone or the tombstone starts at the empty key. If there is no tombstone we'll be skipping the `start_key_` tombstone checks and everything is fine. If the tombstone starts at the empty key we'll skip comparing the tombstone start against the key and assume it covers which is fine unless the user-key comparator is "reversed". But if the user-key comparator is reversed (i.e. `ReverseBytewiseComparator`) the empty key is the largest key and can't be the start tombstone key in any non-empty range tombstone. For example, if we're using `ReverseBytewiseComparator` and an sstable contains:

```
"b"
"a"
""
```

The range tombstone `["a", ""]` will only cover key `"a"`. And the range tombstone `["", ""]` covers nothing.

My brain is wrapped in knots right now. Let me think about this more.

Reply:
In your example with `ReverseBytewiseComparator`, I would've thought the range tombstone `["a", ""]` covers both `"a"` and `""` since `FindKeyForward` thinks all keys to the right are covered when it sees empty end key.

By the way I'll be gone for about a week, but will get back to this soon afterwards.

Reply:
Ah, turns out I was not passing in `file_meta` to `BlockBasedReader::NewIterator` from `TableCache::NewIterator` which was causing the test to pass incorrectly (I think that got lost in a rebase). After fixing that, the test does see both `"a"` and `""` covered highlighting the need to fix `InitRangeTombstone`.

Reply:
Ok, I've got the code all fixed up, though the changes were a bit larger than expected. I need to add some more direct testing of the new code in `BlockBasedTableIterator`. I'll probably have this done before you return.